ui2/src/pages/login/QuickSetup.tsx

```diff
@@ -0,0 +1,560 @@
+  creatingRoleContracts: boolean;
+  onComplete: () => void;
+}) => {
+  const {
+    credentials,
+    serviceSetupData,
+    setServiceSetupData,
+    parties,
+    onComplete,
+    creatingRoleContracts,
+  } = props;
+
+  const [status, setStatus] = useState<string>();
+  const [loading, setLoading] = useState<boolean>(false);
+  const [marketSetupDataMap, setMarketSetupDataMap] = useState<Map<string, string[]>>(new Map());
+
+  const ledger = useLedger();
+  const operator = useWellKnownParties().parties?.userAdminParty || 'Operator';
+
+  const custodianRoles = useStreamQueries(CustodianRole);
+  const exchangeRoles = useStreamQueries(ExchangeRole);
+  const distributorRoles = useStreamQueries(DistributorRole);
+  const settlementServices = useStreamQueries(SettlementService);
+  const matchingServices = useStreamQueries(MarchingService);
+  const operatorService = useStreamQueries(OperatorService);
+
+  const selectedParty = serviceSetupData?.party;
+  const selectedService = serviceSetupData?.service;
+
+  useEffect(() => {
+    setLoading(
+      custodianRoles.loading ||
+        distributorRoles.loading ||
+        settlementServices.loading ||
+        exchangeRoles.loading ||
+        matchingServices.loading ||
+        operatorService.loading
+    );
+  }, [
+    custodianRoles.loading,
+    distributorRoles.loading,
+    settlementServices.loading,
+    exchangeRoles.loading,
+    matchingServices.loading,
+    operatorService.loading,
+  ]);
+
+  useEffect(() => {
+    if (loading) {
+      return;
+    }
+
+    let newMarketData: Map<string, string[]> = new Map();
+
+    custodianRoles.contracts.forEach(c => addMarketData(c.payload.provider, 'Custody'));
+    distributorRoles.contracts.forEach(c => addMarketData(c.payload.provider, 'Listing'));
+    settlementServices.contracts.forEach(c => addMarketData(c.payload.provider, 'Settlement'));
+    exchangeRoles.contracts.forEach(c => addMarketData(c.payload.provider, 'Trading'));
+    matchingServices.contracts.forEach(c => addMarketData(c.payload.provider, 'Matching'));
+
+    function addMarketData(party: string, serviceName: string) {
+      newMarketData.set(party, [...(newMarketData.get(party) || []), serviceName]);
+    }
+
+    setMarketSetupDataMap(newMarketData);
+  }, [
+    loading,
+    custodianRoles.contracts.length,
+    distributorRoles.contracts.length,
+    settlementServices.contracts.length,
+    exchangeRoles.contracts.length,
+    matchingServices.contracts.length,
+  ]);
+
+  useEffect(() => {
+    const createOperatorService = async () => {
+      await ledger.create(OperatorService, { operator: credentials.party });
+    };
+
+    if (!operatorService.loading && operatorService.contracts.length === 0) {
+      createOperatorService();
+    }
+  }, [operatorService.loading, operatorService]);
+
+  const partyOptions = parties.map(party => {
+    return { text: party.partyName, value: party.party };
+  });
+
+  let serviceOptions = services.map(service => {
```

Comment:
you can map over the enum directly, and avoid having to maintain two different vars with service lists in them.

`Object.values(ServiceKind).map(...)` should work, (after importing ServiceKind -- see above), and now we don't need the `services` array variable

Reply:
I was going to use this trick for `MarketRole` (the daml sum type enum) for something but the Daml codegen adds a few extra entries related to the deserialization so it wouldn't have worked 😞